modules/ml/src/main/java/org/apache/ignite/ml/dataset/DatasetBuilder.java

```diff
@@ -48,6 +50,16 @@
     public <C extends Serializable, D extends AutoCloseable> Dataset<C, D> build(
         PartitionContextBuilder<K, V, C> partCtxBuilder, PartitionDataBuilder<K, V, C, D> partDataBuilder);
 
+    /**
+     * Get upstream transformers chain. This chain is applied to upstream data before it is passed
+     * to {@link PartitionDataBuilder} and {@link PartitionContextBuilder}. This is needed to allow
+     * transformation to upstream data which are agnostic of any changes that happen after.
+     * Such transformations may be used for deriving meta-algorithms such as bagging
+     * (see {@link BaggingUpstreamTransformer}).
+     *
+     * @return Upstream transformers chain.
+     */
+    public UpstreamTransformerChain<K, V> upstreamTransformersChain();
```

Comment:
Honestly speaking this method looks weird here. This class is a builder, so I understand `build` method. Okay, it's modifiable, so we have `withFilter` method. But why do we have a getter method here?

If you think we need to update this builder by adding new transformers it would be better ho have a `withTransformer` method. What do you think?

Reply:
This would be ok of we do not need access to the chain itself, which in turn we need to assign seed to it.

Reply:
And `DatasetBuilder` user should know if the builder encapsulates transformers or not? And if yes user should know that transformers use some seed (or maybe they don't use seed, who knows) in work and that seed should be manually set up? Sorry, but in this case it looks like a very leaky abstraction.